src/main/java/com/hazelcast/aws/impl/AwsOperation.java

```diff
@@ -0,0 +1,332 @@
+        this.awsCredentials = new AwsCredentials(awsConfig);
+        this.endpointURL = endpointURL;
+        this.service = service;
+        this.docVersion = docVersion;
+        this.httpMethod = httpMethod;
+    }
+
+    protected boolean isEcsEndpoint() {
+        return endpointURL.getAuthority().startsWith("ecs.");
+    }
+
+    /**
+     * Invokes this service, unmarshal the response and return it.
+     *
+     * @return the response
+     * @throws Exception if there is an exception invoking the service
+     */
+    public E execute(Object... args) throws Exception {
+        if (isNotEmpty(awsCredentials.getIamRole()) || isEmpty(awsCredentials.getAccessKey())) {
+            retrieveCredentials();
+        }
+
+        LOGGER.finest("OK we have credentials, signing request...");
+
+        prepareHttpRequest(args);
+
+        Aws4RequestSigner requestSigner = getRequestSigner();
+        requestSigner.sign(attributes, headers, body, httpMethod);
+        headers.put("Authorization", requestSigner.getAuthorizationHeader());
+        String securityToken = awsCredentials.getSecurityToken();
+        if (StringUtil.isNotEmpty(securityToken)) {
+            headers.put("X-Amz-Security-Token", securityToken);
+        }
+
+        InputStream stream = null;
+
+        try {
+            LOGGER.finest("Calling service at " + endpointURL.toExternalForm());
+            stream = callServiceWithRetries();
+            return unmarshal(stream);
+        } finally {
+            closeResource(stream);
+        }
+    }
+
+    /**
+     * @param stream
+     * @return
+     */
+    abstract E unmarshal(InputStream stream);
+
+    /**
+     * @param args
+     */
+    protected void prepareHttpRequest(Object... args) {
+        body = "";
+    }
+
+    /**
+     * Add available filters to narrow down the scope of the query
+     * Override where appropriate.
+     */
+    protected void addFilters() {
+    }
+
+    /**
+     * AWS response codes for client and server errors are specified here:
+     * {@see http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html}.
+     */
+    private static boolean isAwsError(int responseCode) {
+        return responseCode >= MIN_HTTP_CODE_FOR_AWS_ERROR && responseCode < MAX_HTTP_CODE_FOR_AWS_ERROR;
+    }
+
+    private static String extractErrorMessage(HttpURLConnection httpConnection) {
+        InputStream errorStream = httpConnection.getErrorStream();
+        if (errorStream == null) {
+            return "";
+        }
+        return readFrom(errorStream);
+    }
+
+    private static String readFrom(InputStream stream) {
+        Scanner scanner = new Scanner(stream, UTF8_ENCODING).useDelimiter("\\A");
+        return scanner.hasNext() ? scanner.next() : "";
+    }
+
+    // Visible for testing
+    void retrieveCredentials()
+            throws IOException {
+        if (isEcsEndpoint()) {
+            retrieveContainerCredentials(getEnvironment());
```

Comment:
I think credential retrieval part should not be in AWSOperation or we should have separate ECSOperation and EC2Operation. Having an if statement is already a bad smell...

Reply:
Agreed, I will take care of this